middleware/proxy.go

```diff
@@ -287,19 +302,52 @@ func ProxyWithConfig(config ProxyConfig) echo.MiddlewareFunc {
 				req.Header.Set(echo.HeaderXForwardedFor, c.RealIP())
 			}
 
-			// Proxy
-			switch {
-			case c.IsWebSocket():
-				proxyRaw(tgt, c).ServeHTTP(res, req)
-			case req.Header.Get(echo.HeaderAccept) == "text/event-stream":
-			default:
-				proxyHTTP(tgt, c, config).ServeHTTP(res, req)
-			}
-			if e, ok := c.Get("_error").(error); ok {
-				err = e
-			}
+			retries := config.RetryCount
+			for {
+				var tgt *ProxyTarget
+				var err error
+				if isTargetProvider {
+					tgt, err = provider.NextTarget(c)
+					if err != nil {
+						return err
+					}
+				} else {
+					tgt = config.Balancer.Next(c)
+				}
+				c.Set(config.ContextKey, tgt)
+
+				// Proxy
+				switch {
+				case c.IsWebSocket():
+					proxyRaw(tgt, c).ServeHTTP(res, req)
+				case req.Header.Get(echo.HeaderAccept) == "text/event-stream":
+				default:
+					proxyHTTP(tgt, c, config).ServeHTTP(res, req)
+				}
 
-			return
+				e, hasError := c.Get("_error").(error)
+				if !hasError {
+					return nil
+				}
+
+				retry := false
+				if httpErr, ok := e.(*echo.HTTPError); ok {
```

Comment:
I did remove this check with the intention of leaving it to the filter function to decide, but after looking at the rest of the proxy code, I couldn't see any situation where another type of error would want to be retried. I decided to keep this check, so they user provided function can be simplified.